File: Data_Structure/Java/LinkedListReverse.java
class NodeR {
    String data;
    NodeR next;

    NodeR(String data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedListR{
    private NodeR head;
    LinkedListR() {
        this.head = null;
    }
    public void insertAtBeginning(String data){
        NodeR newNode = new NodeR(data);
        if (head == null){
            head = new NodeR(data);
            return;
        }else{
            newNode.next= head;
            head = newNode;
        }
    }
    public void insertMultipleValues(String[] data_list){
        for (String value : data_list){
            insertAtBeginning(value);
        }
    }
    public void traverse(){
        NodeR current = head;
        if(current == null){
            System.out.println("List is empty");
            return;
        }
        while(current != null){
            System.out.print(current.data + " -> ");
            current = current.next;
        }
    }
    public void reverseList(){
        NodeR prev = null;
        NodeR current = head;
        NodeR next = null;
        while(current!=null){
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
    }
}

public class LinkedListReverse {
    public static void main(String[] args) {
        LinkedListR list = new LinkedListR();
        list.insertAtBeginning("A");
        list.insertAtBeginning("B");
        list.insertMultipleValues(new String[]{"Apple", "Banana", "Cherry","Date"});
        list.traverse();
        System.out.println("\nReversed List:");
        list.reverseList();
        list.traverse();

    }
}

File: Data_Structure/Python/LinkedListReverse.py
class Node:
    def __init__(self,data=None,next=None):
        self.data=data
        self.next=next
class LinkedList:
    def __init__(self):
        self.head = None
    def insertAtBegining(self, data):
        node =Node(data,self.head)
        if self.head is None:
            self.head = node
            return
        node.next = self.head
        self.head = node
    def insertMultipleValues(self,data_list):
        for data in data_list:
            self.insertAtBegining(data)
    def traverse(self):
        if self.head is None:
            print("The linked list is empty")
            return
        itr = self.head
        linkedListStr = ""
        while itr:
            linkedListStr += str(itr.data) + "-->"
            itr = itr.next
        print(linkedListStr)

    def reverse(self):
        if self.head is None:
            print("The linked list is empty")
            return
        prev = None
        current = self.head
        while current is not None:
            next = current.next #or alternative next = next.next
            current.next = prev
            prev = current
            current = next
        self.head = prev
        

ll=LinkedList()
ll.insertAtBegining(1)
list = ["Apple", "Banana", "Cherry", "Date", "Elderberry"]
ll.insertMultipleValues(list)
ll.traverse()
ll.reverse()
ll.traverse()